### PR TITLE
[Dev] Add batched gemm for `DSv4HybridAttention`

### PR DESCRIPTION
**Pull request by yaox12**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?

Use `transformer_engine.pytorch.modules.BatchedLinear` for `DSv4HybridAttention.linear_o_group_proj`, which 
1. supports MXFP8 and BF16
2. eliminates the `.contiguous()` call after the group projection.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
3. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

File: megatron/core/transformer/experimental_attention_variant/deepseek_v4_hybrid_attention.py [hunk 1]
```diff
@@ -219,6 +232,108 @@ def __init__(
...
+        op = self._linear_o_group_proj_op[0]
+        if op.weight is not self.linear_o_group_proj:
+            # Parameter wrappers may replace the registered Parameter after init.
+            op.weight = self.linear_o_group_proj
+        return op
+
+    def _apply_linear_o_group_proj(self, input_: torch.Tensor) -> torch.Tensor:
+        """Apply TE BatchedGEMM when available, otherwise use the legacy einsum."""
+        op = self._get_linear_o_group_proj_op()
+        if op is None:
```
Comment (hxbai) [inline megatron/core/transformer/experimental_attention_variant/deepseek_v4_hybrid_attention.py]: Can we add a `linear_o_group_proj` to `DSv4HybridSelfAttentionSubmodules` spec to select the TE op?

I think the silent fallback makes it not easy to locate some issues.

File: megatron/core/transformer/experimental_attention_variant/deepseek_v4_hybrid_attention.py [hunk 2]
```diff
@@ -219,6 +263,149 @@ def __init__(
...
+            global_shape=checkpoint_global_shape,
+            global_offset=checkpoint_global_offset,
+            axis_fragmentations=checkpoint_axis_fragmentations,
+        )
+
+        @torch.no_grad()
+        def build_fn(
+            key: str, tensor: torch.Tensor, replica_id: ReplicaId, flattened_range: Optional[slice]
+        ) -> ShardedTensor:
+            checkpoint_tensor = tensor.reshape(checkpoint_local_shape)
```
[the remark this answers is not in the document]
Comment (yaox12) [reply megatron/core/transformer/experimental_attention_variant/deepseek_v4_hybrid_attention.py]: > The previous implementation did assume full-shaped factory data, so it would fail if invoked with a flat partial optimizer state.
>
> After tracing the checkpoint entry point, I changed the fix slightly instead of adding DSv4-specific support for `fully_sharded_model_space`:
>
> - `fully_sharded_model_space` is only the fallback when `DistributedOptimizer.sharded_state_dict()` is called without checkpoint metadata.
> - The normal checkpoint path sets `dp_reshardable` by default, or `fully_reshardable` when requested. `dp_reshardable` does not invoke this parameter factory, while `fully_reshardable` passes full-shaped state with `flattened_range=None`.
> - `fully_sharded_model_space` is deprecated and is already incompatible globally because `ShardedTensor.validate_metadata_integrity()` rejects every non-`None` `flattened_range`. Supporting it only for this DSv4 parameter would therefore not make a complete optimizer checkpoint valid.
>
> I simplified the implementation accordingly:
>
> - The fallback path now retains the original 2D `Parameter`, so it uses the existing state-dict and distributed-checkpoint behavior directly.
> - Only the TE `BatchedLinear` path installs the 3D-to-2D compatibility factory.
> - The factory supports the active non-flat checkpoint formats, preserves PP/FSDP group-axis metadata, and explicitly rejects the deprecated flat format.
>
> I also added coverage for the fallback 2D `ShardedTensor` path and pipeline-offset metadata. The full test file passes on GB200 with two ranks (`35 passed` on each rank), including BF16, MXFP8, quantized model initialization, legacy checkpoint loading, and the einsum fallback.
>
> If `fully_sharded_model_space` needs to be restored, I think it should be addressed globally in the distributed-checkpoint/optimizer implementation rather than with a DSv4-specific irregular-shard conversion.

**Review by hxbai — APPROVED** [review]

**Comment by hxbai** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/pull/5722/commits/cbbd9ea92232a75cc77df3a58ff971b179c3efd4
